pkg/binding/binding.go

```diff
@@ -51,35 +53,81 @@ func (c *Controller) syncBinding(ctx context.Context, bindingName string) error
 			ObjectMeta: metav1.ObjectMeta{
 				Name: bindingName},
 		}
-	} else if err != nil {
-		return fmt.Errorf("failed to get Binding from informer cache (name=%v): %w", bindingName, err)
+	} else if bindingErr != nil {
+		return fmt.Errorf("failed to get Binding from informer cache (name=%v): %w", bindingName, bindingErr)
 	}
 
 	// binding name matches that of the bindingpolicy 1:1, therefore its NamespacedName is the same.
 	bindingPolicyIdentifier := binding.GetName()
 
+	policy, policyErr := c.bindingPolicyLister.Get(bindingPolicyIdentifier)
+	// `*policy` is immutable
+	if errors.IsNotFound(policyErr) {
+		logger.V(2).Info("Aborting sync of Binding because the corresponding Policy is gone", "name", bindingName)
+		return nil
+	} else if policyErr != nil {
+		return fmt.Errorf("failed to get BindingPolicy from informer cache (name=%v): %w", bindingName, policyErr)
+	}
+
 	// generate binding spec from resolver
 	generatedBindingSpec := c.bindingPolicyResolver.GenerateBinding(bindingPolicyIdentifier)
 	if generatedBindingSpec == nil { // resolution does not exist, abort syncing
 		return fmt.Errorf("syncing Binding was stopped because it has no counterpart resolution")
 	}
 
 	// calculate if the resolved decision is different from the current one
-	if !c.bindingPolicyResolver.CompareBinding(bindingPolicyIdentifier, &binding.Spec) {
+	if c.bindingPolicyResolver.CompareBinding(bindingPolicyIdentifier, &binding.Spec) {
+		logger.V(4).Info("Binding is up to date", "name", binding.GetName())
+	} else {
 		// update the binding object in the cluster by updating spec
-		if err = c.updateOrCreateBinding(ctx, binding, generatedBindingSpec); err != nil {
+		if err := c.updateOrCreateBinding(ctx, binding, generatedBindingSpec); err != nil {
 			return fmt.Errorf("failed to update or create binding: %w", err)
 		}
 
 		// notify the bindingpolicy resolution broker that the binding has been updated
 		c.bindingPolicyResolver.Broker().NotifyCallbacks(bindingPolicyIdentifier)
-		return nil
+	}
+	srPerObj := c.bindingPolicyResolver.GetSingletonReportedStateRequestsForBinding(bindingPolicyIdentifier)
+	policyErrors := append([]string{}, binding.Status.Errors...)
+	badSR := []objectWithNumWECs{}
+	for _, srStatus := range srPerObj {
+		if srStatus.WantSingletonReportedState && srStatus.NumWECs != 1 {
```

Comment:
Similar to the pervious comment on the code comments - what do we think the user expects  to see in the case of 0 wecs ?

Reply:
The singleton reported state return feature only makes sense to use when the number of associated WECs is exactly 1. When the number is 0, there will be no return of reported state into the workload object; but since the user requested such return, there is a surprise here that should be reported.